vhdl/quartus/quartus_workflow.py

```diff
@@ -0,0 +1,263 @@
+    log_file_path = working_dir + "qsys_script.log"
+    log_msg = "Generating qsys file from tcl file"
+    run_cmd_and_log(cmd, log_msg, log_file_path)
+
+def gen_qsys_system(qsys_file, working_dir):
+    """Generates the qsys file using qsys-generate
+
+    Parameters
+    ----------
+    qsys_file : string
+        Qsys file to generate
+    working_dir : string
+        Working directory of the generation process
+    """
+    cmd = f'cd {working_dir} & qsys-generate --synthesis=VHDL --search-path=\"..\\component_library\\**\\*,$\"  {qsys_file}'
+    log_file_path = working_dir + "qsys_gen.log"
+    log_msg = "Generating system"
+    run_cmd_and_log(cmd, log_msg, log_file_path)
+
+def gen_project_tcl(target, template, working_dir):
+    """Generates the make_project.tcl file
+    Also copies over the top level VHDL file and
+    base project tcl file into the working directory
+
+    Parameters
+    ----------
+    target : Target
+        NamedTuple that provides information about the target configuration
+    template : qsys_templates
+        template object to utilize qsys templates  
+    working_dir : string
+        Working directory of the generation process
+    """
+
+    base_project_file = target.base_proj_tcl_file
+    top_level_vhdl_file = target.top_level_vhdl_file
+    original_system = target.original_system
+
+    print("Generating make_project.tcl")
+    with open(working_dir + "make_project.tcl", "w") as proj_file:
+        proj_file.write(template.add_quartus_project(target.system_name, target))
+    copyfile(RES_DIR + base_project_file, working_dir + base_project_file)
+
+    with open(RES_DIR + top_level_vhdl_file, "r") as orig_top_file:
+        with open(working_dir + top_level_vhdl_file, "w") as new_top_file:
+            for line in orig_top_file:
+                new_top_file.write( line.replace( original_system, target.system_name ) )
+
+def gen_pll_qsys(working_dir):
+    """Generates the phase locked loop component needed for the arria10 configuration
+
+    Parameters
+    ----------
+    working_dir : string
+        Working directory of the generation process
+    """
+    pll_file = "pll.qsys"
+    copyfile(RES_DIR + "\\res\\" + pll_file, working_dir + pll_file)
+    log_msg = "Generating pll"
+    cmd =  f'cd {working_dir} & qsys-generate --synthesis=VHDL --search-path=\"..\\component_library\\**\\*,$\"  {pll_file}'
+    log_file_path = working_dir + "pll_gen.log"
+    
+    run_cmd_and_log(cmd, log_msg, log_file_path)
+
+def gen_project(working_dir):
+    """Generates and compiles the project defined in make_project.tcl
+
+    Parameters
+    ----------
+    working_dir : string
+        Working directory of the generation process
+    """
+    log_msg = "Generating and compiling project"
+    cmd = f"cd {working_dir} & quartus_sh -t make_project.tcl"
+    log_file_path = working_dir + "project_gen.log"
+    
+    run_cmd_and_log(cmd, log_msg, log_file_path)
+
+def gen_rbf(working_dir, target_system):
```

Comment:
I'd prefer to hook the rbf generation into the quartus project itself as a post-compile hook, then we always get an rbf every time we compile a project, even if we aren't using this autogen script. We are planning on doing the same with the device tree overlays. 

We can worry about that decision during the full autogen code review

Reply:
It would probably be better to make that change now rather than later.  Don't want things to keep piling up. ;)

Reply:
Fair enough. We will want to make the rbf creation a post-compile hook for every project, so I didn't want to hold up this code review as it's more about simply creating a Quartus project. But I agree that if we have time, we might as well change it now, or at least start. I think the full codebase review will potentially have implications on how we organize our projects, post-compile hooks, etc., though, so I see some value in waiting to get that sorted before changing all of the Quartus projects. 

I also think getting the device tree generation stuff wrapped up this cycle is higher priority. Either way, good point ;)

Reply:
I back off on my previous arguments a bit. Creating the post-compile hook here would be a matter of creating another tcl script and adding some lines to the *_proj.tcl files. That is separate from converting all of our existing projects to using the post-compile hook. 

So yes, if we all agree on using a post-compile hook instead, then let's do it here now 😉

Reply:
Yeah, let's go ahead and do it.  I figured it would be pretty straightforward.  Even though it would fit in the review for the whole process, fixing it sooner rather than later would save time in the long run.

Reply:
Here's some [documentation](https://www.intel.com/content/www/us/en/programmable/support/support-resources/design-examples/design-software/tcl/auto_processing.html) on adding tcl hooks into the compilation process. There are also some examples in our [Dropbox swap folder](https://www.dropbox.com/home/FlatEarth%20Team%20Folder/Swap/NIH/quartus%20autogen%20hooks/scripts)

Reply:
Done. The current implementation, which exists in`res/gen_rbf.tcl`, loops through all sof files found in the output_files directory and converts them to rbf files. This keeps it flexible as it doesn't need to know the name of the sof file(s) to convert and isn't going to run into issues if there are multiple in the directory for whatever reason. It is also quick, so running it more than needed isn't a significant issue